app/uk/gov/hmrc/transitmovements/services/MessagesXmlParsingService.scala

```diff
@@ -53,37 +56,57 @@ class MessagesXmlParsingServiceImpl @Inject() (implicit materializer: Materializ
   // we don't want to starve the Play pool
   implicit val ec: ExecutionContext = ExecutionContext.fromExecutorService(Executors.newCachedThreadPool())
 
-  private def buildMessageData(messageTypeMaybe: ParseResult[MessageType], dateMaybe: ParseResult[OffsetDateTime]): ParseResult[MessageData] =
-    for {
-      messageType    <- messageTypeMaybe
-      generationDate <- dateMaybe
-    } yield MessageData(messageType, generationDate)
+  def buildMessageData(
+    dateMaybe: ParseResult[OffsetDateTime],
+    mrnMaybe: ParseResult[MovementReferenceNumber],
+    messageType: MessageType
+  ): ParseResult[MessageData] =
+    if (messageType == MrnAllocated)
+      for {
+        generationDate <- dateMaybe
+        mrn            <- mrnMaybe
+      } yield MessageData(generationDate, Some(mrn))
+    else
+      dateMaybe.map(
+        generationDate => MessageData(generationDate, None)
+      )
 
-  private val messageFlow: Flow[ByteString, ParseResult[MessageData], NotUsed] =
-    Flow.fromGraph(
-      GraphDSL.create() {
-        implicit builder =>
-          import GraphDSL.Implicits._
+  def messageFlow(messageType: MessageType) = Flow.fromGraph(
+    GraphDSL.create() {
+      implicit builder =>
+        import GraphDSL.Implicits._
 
-          val broadcast = builder.add(Broadcast[ParseEvent](2))
-          val combiner  = builder.add(ZipWith[ParseResult[MessageType], ParseResult[OffsetDateTime], ParseResult[MessageData]](buildMessageData))
+        val broadcastXml = builder.add(Broadcast[ParseEvent](2))
+        val combiner =
+          builder.add(
+            ZipWith[ParseResult[OffsetDateTime], ParseResult[MovementReferenceNumber], ParseResult[MessageData]](
+              (date, mrn) => buildMessageData(date, mrn, messageType)
+            )
+          )
 
-          val xmlParsing      = builder.add(XmlParsing.parser)
-          val messageTypeFlow = builder.add(XmlParsers.messageTypeExtractor)
-          val dateFlow        = builder.add(XmlParsers.preparationDateTimeExtractor)
+        val xmlParsing = builder.add(XmlParsing.parser)
+        val mrnFlow    = builder.add(XmlParsers.movementReferenceNumberExtractor)
+        val dateFlow   = builder.add(XmlParsers.preparationDateTimeExtractor(messageType))
+        val partition  = builder.add(Partition[ParseEvent](2, _ => if (messageType == MrnAllocated) 0 else 1))
```

Comment:
This feels a little hacky and brittle - and I wonder if there is a better way to do this. If the extractor itself returns nothing (and we let it, rather than throw an exception), can we just build the message data based on the optionals/maybes that are returned?